### ISSUE TITLE
[selectors] Add pseudo-classes for `<select>` being a drop-down box vs a list box

### ISSUE THREAD

**Issue by zcorpan** [w3c/csswg-drafts#7422]
See https://github.com/whatwg/html/issues/7050

Chromium [supports](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/resources/html.css;l=909?q=html.css&ss=chromium) a `:-internal-list-box` pseudo-class in the UA stylesheet, which matches [`<select>` elements that render as a list box instead of as a dropdown box](https://html.spec.whatwg.org/multipage/rendering.html#the-select-element-2).

Gecko similarly [supports](https://searchfox.org/mozilla-central/source/layout/style/res/forms.css#249) `:-moz-select-list-box` in the UA stylesheet.

WebKit [seems to use attribute selectors instead](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/css/html.css#L1060), although I think it's not entirely accurate since [the decision per spec](https://html.spec.whatwg.org/multipage/rendering.html#list-box) depends on the [parsing result of the `size` attribute](https://html.spec.whatwg.org/multipage/form-elements.html#concept-select-size). (Demo: `<select size="+02xyz">`)

Web developers could rely on the `multiple` attribute being present to target list boxes, but that might not always work so well, in particular if you're writing CSS that should work for arbitrary HTML.

I think it makes sense to add a pseudo-class that matches `select` elements that are list boxes, e.g. `:select-list-box`. To match drop-downs, you can do `select:not(:select-list-box)`.

Spec: https://drafts.csswg.org/selectors/

**Comment by zcorpan** [comment]
Is there interest in standardizing this pseudo-class (and make it also available to author CSS)?

cc @mfreed7 

Via git blame of WebKit's html.css I found https://bugs.webkit.org/show_bug.cgi?id=220376 -- I think a pseudo-class as suggested here would be less prone to bugs like this (compared to attribute selectors). cc @pxlcoder

Via git blame of Gecko's html.css I found https://bugzilla.mozilla.org/show_bug.cgi?id=1643246 where attribute selectors were replaced with a pseudo-class. cc @emilio @heycam

**Comment by mfreed7** [comment]
It certainly seems tricky to select this based on attribute selectors, so I'm supportive of trying to standardize a pseudo class.

Picking a name will be tricky. I'm not sure I like `:select-list-box`, since in my view, both in-page and pop-up `<select>`s contain a "listbox". One is just a pop-up listbox and the other is an in-page listbox. Perhaps something like `:select-pops-up`? Or at least something that refers to the pop-up behavior? Similarly, we need to be careful to distinguish whether this pseudo class refers to whether the listbox pops up or not (as I've assumed), or whether it refers to whether multi-select is supported. Those are (at least theoretically) distinct.

**Comment by bradkemper** [comment]
I believe the `select` element has an aria list box role either way, so that could be confusing.

**Comment by zcorpan** [comment]
How about "list box" and "dropdown list box"? (Not sure about "list box" vs "listbox".)

**Comment by bradkemper** [comment]
"Drop down" or "drop down list box" seem pretty clear (even though they are "pop up menus" on Mac). But just "list box" by itself is not clear, because I honestly wouldn't remember which one you meant without comparing it to the other term. I think if you had "multiple list box" (or similar*) and "drop down list box" as the two terms, then they'd both be clear, even in isolation. 

* similar:

- List box multiple 
- Multi-select list box
- no-drop list box (clear, but ugh)

**Comment by zcorpan** [comment]
It's not necessarily multi-select, though.

```html
<select size=5>
  <option>this is still single-select list box
  <option>foo
  <option>bar
</select>
```

**Comment by zcorpan** [comment]
We could define only a pseudo-class for the drop-down, and rely on the `:not()` function to match the not-drop-down state.

```css
select:drop-down-list-box {}
select:not(:drop-down-list-box) {}
```

**Comment by dbaron** [comment]
+1 to the idea of targeting the idea of having a dropdown.  Perhaps something like `:has-dropdown`?

**Comment by bradkemper** [comment]
@zcorpan works for me. I also like the @dbaron refinement. Though there should probably be another dash, like `:has-drop-down`, so we don’t repeat the nowrap mistake. 

Also, would a native date picker or color picker be considered a drop-down, like `input[type=date]:has-drop-down`?

**Comment by bradkemper** [comment]
I’m ambivalent about the extra dash, really.

**Comment by zcorpan** [comment]
> Also, would a native date picker or color picker be considered a drop-down, like `input[type=date]:has-drop-down`?

Good question. I would say no. The date widgets are not guaranteed to have a drop-down or a popup, it would be conforming to have a control with only spinners, for example. The color well is [specified](https://html.spec.whatwg.org/multipage/rendering.html#the-input-element-as-a-colour-well) to have a "color picker" that is separate from the color well, but nothing says it has to be a drop-down. Firefox on macOS opens a popup window for the color picker.

Since the need is specific to the `select` element, I think we should define it to only apply to `select`. Should the naming be different because of that?

**Comment by mfreed7** [comment]
It seems to me that if this is only ever going to apply to `<select>`, then it should have "select" in the name. Otherwise, this feels confusing:

```css
input[type=date]:has-dropdown { doesn't match, even when the date picker "drop down menu" is present }
```

As for `:has-dropdown` and `:no-dropdown` - isn't that exactly why we have `:not()`, to avoid the need for that?

I vote for `:select-with-dropdown`, because "with" indicates the `<select>` comes with a drop-down. Somehow (and I'm not sure why) but `:select-has-dropdown` sounds to me like it should apply only when the dropdown is open. Like it **currently** has a dropdown.

Interesting side reading on "drop-down" vs "dropdown": https://www.chicagomanualofstyle.org/qanda/data/faq/topics/HyphensEnDashesEmDashes/faq0153.html. Perhaps we should use "menu" instead?

**Comment by zcorpan** [comment]
I don't think "menu" distinguishes `<select>` vs `<select size=5>`. The name here is also "for developers", so we're good per the Microsoft Writing Style Guide. :-)

`:not()` indeed allows the same expressiveness, but it's a bit more cumbersome compared to using a direct pseudo-class. Compare with

```
:link { ... }
:visited { ... }
```
vs
```
:any-link:not(:visited) { ... }
:visited { ... }
```

**Comment by tabatkins** [comment]
+1 to "put `select` in the name" - it means we don't have to figure out whether this also applies to N other things, and verify if new things need to have it, possibly in only certain UAs, etc... Plus the general argument that if it only applies to 1 thing now, people *will* rely on that fact and use it on its own (without a tagname selector) with the intention of targeting `select`s specifically, and that code will then break in the future when we apply it to more. General policy: never use a generic name that *could plausibly* apply to many things unless there are *at least* three things that it *already* applies to.

Relying solely on `:not()` for the non-dropdown case is slightly unfortunate, as it requires you to use the tagname in the selector as well - `select:not(:select-with-dropdown)` - since just `:not(:select-with-dropdown)` will target random elements too. I think we should come up with a name for the other state. I think `listbox` is fine - even if it's a little unclear on its own, since it'll be contrasted with `dropdown` I think it's reasonable.

**Comment by zcorpan** [comment]
Since the dropdown menu is called a listbox in [APG](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/) and [Open UI](https://open-ui.org/prototypes/selectmenu), it's plausible people will assume `listbox` means the dropdown menu.

The `<select>` element certainly will not get new widget states. New widgets should be new elements.

**Comment by zcorpan** [comment]
So it sounds like `:select-with-dropdown` is the winner for the "with dropdown" state.

For the other state, we have:

- `:select-no-dropdown`
- `:select-without-dropdown`

In https://drafts.csswg.org/indexes/ I see several "no-" but no "without-", so the former seems more CSS-y and is also shorter.

Thus, I suggest we go with `:select-with-dropdown` and `:select-no-dropdown`.

**Comment by bramus** [comment]
> In https://drafts.csswg.org/indexes/ I see several "no-" but no "without-", so the former seems more CSS-y and is also shorter.

Looking at that list, I see that these `no-` properties/values are (mostly) prefixed version of other ones that do not have a prefix. For example `no-clip` is the antonym of `clip`; not `has-clip` or `with-clip`.

Working my way back from the suggested `select-no-dropdown`, it should go hand in hand with `select-dropdown` then?

If `select-dropdown` does not make sense, and we do want to keep `select-with-dropdown`, its antonym – in similar fashion as `clip`/`no-clip` – should then become `select-no-with-dropdown`? From a linguistic POV the latter doesn’t make sense at all, and it would become `select-without-dropdown`.

tl;dr If we have `select-with-dropdown`, `select-without-dropdown` seems like a better opposite to me.

**Comment by nt1m** [comment]
Might be good to reuse the well known CSS `appearance` values, as vocabulary:

`select:menulist` and `select:listbox`

**Comment by zcorpan** [comment]
I think only `:listbox` is not great because there are other listboxes other than "`<select>` as a listbox" (for example, the dropdown part of a dropdown `<select>` is also called a listbox). The `appearance` values are also legacy in favor of `auto`.

**Comment by tabatkins** [comment]
Yeah, and I wouldn't really consider them "well-known" either. I couldn't tell you what `appearance` values exist, for example, besides `auto` and `button`.

**Comment by zcorpan** [comment]
I should be able to remotely attend the TPAC CSS meeting either Thursday or Friday, but only before lunch. cc @astearns

**Comment by bramus** [comment]
Sidenote: If authors had [the ability to numerically compare attribute values](https://github.com/w3c/csswg-drafts/issues/5890), they would be able to select those elements using this snippet:

```css
select[size>1],
select[multiple] {
  /*🎉 */
}
```

This, in turn, would make a lovely [custom selector](https://drafts.csswg.org/css-extensions-1/#custom-selectors) as well:

```css
@custom-selector :--with-listbox [size>1], [multiple];

select:--with-listbox {
  /*🎉 */
}
```

**Comment by zcorpan** [comment]
@bramus not sufficient for the UA stylesheet, at least, since browsers need to get cases like `<select size="+02xyz">` correct.

**Comment by zcorpan** [comment]
@astearns is it possible to move this issue to some time before lunch (either Thursday or Friday)?

**Comment by lukewarlow** [comment]
Based on the fact we now have `::picker(select)` which exists for dropdown select but not for in-page select. Could we do

`select:has-picker(select)` as the new selector? Question would be does this match in non-base appearance mode?

If it does match then `select:not(:has-picker(select))` works for listbox. If it doesn't work then that's trickier. You'd have to do listbox styles and then potentially undo them?

**Comment by lukewarlow** [comment]
Adding this to the agenda as it would be useful to resolve on something. This is a soft blocker for base appearance listboxes.

**Comment by lukewarlow** [comment]
@nt1m should we try and get this on the agenda for next week? It'd be handy to get a resolution on this to unblock #12468

**Comment by nt1m** [comment]
> [@nt1m](https://github.com/nt1m) should we try and get this on the agenda for next week? It'd be handy to get a resolution on this to unblock [#12468](https://github.com/w3c/csswg-drafts/issues/12468)

Fine by me

**Comment by LeaVerou** [comment]
One thing to note is that adding such a pseudo-class will effectively prevent us from adding CSS to control the listbox vs dropdown mode. That may be fine, but we need to be aware of the tradeoff we are making.

**Comment by josepharhar** [comment]
I don't think that controlling drop-down box vs list box should be controllable via css.

I think that `:has-picker(select)` or `:select-has-picker` both sound reasonable to me.

**Comment by LeaVerou** [comment]
One thing that was brought up in the unminuted part of the WHATWG call last week (I think by @lukewarlow?) was that for all controls with pickers, there may be value in providing a way to render the picker inline. E.g. for date pickers this could be part of a range input control. Perhaps this could also be useful here.

**Comment by LeaVerou** [comment]
To sum up my position from the call, while it seems like useful functionality, the way it is currently discussed has several warts. Using a plain `:has-picker` would imply that this works for other controls with pickers (date inputs, color pickers etc — those targeted by `input.showPicker()`), but `:has-picker(select)` is _incredibly_ weird and unprecedented.

But most importantly, if `:has-picker` only works for `appearance: base` as was discussed, that introduces a circularity:

```css
select {
	appearance: base-select;
}

select:has-picker {
	appearance: auto; /* now it no longer matches */
}
```

We have historically avoided making selector matching depend on CSS properties for this reason.

**Comment by lukewarlow** [comment]
I still feel like this would be an incredibly useful addition, especially when some browsers now render different attribute combos in different modes. I don't personally think a CSS way to control this is desirable, and I think the picker embedded in page cases can be solved via a HTML attribute or something.

The issue of cycles does somewhat still stand. Especially (for example) if Safari has `<select multiple size="1">` render one way in appearance auto and another in appearance base.

But we already have situations where appearance can change in a selector with no effect. The below would potentially even cause cycles if it did work (the checkbox would collpase to 0width and hover would no longer apply). So is this a solved problem?

Ideally this new selector (whatever we call it) would work in all appearance modes.

```
<input type=checkbox>
<style>
input:hover {
    appearance: none;
}
</style>
```

Personally I'd be fine with just a new `:list-box` (name TBD) selector for select specifically for now. If we did ever want to do a :has-picker in future I think that's fine and isn't necessarily even duplicative.
